### PR TITLE
PrimitiveBatch

### PR DESCRIPTION
**Pull request by LithiumToast**
Don't merge it yet. This PR is just for discussion while I fill in the remaining code.

### CODE REVIEW

File: Source/Demos/Demo.Particles/Game1.cs [hunk 1]
```diff
@@ -28,6 +28,7 @@ public Game1()
             Content.RootDirectory = "Content";
             IsMouseVisible = true;
             Window.AllowUserResizing = true;
+            Window.Position = Point.Zero;
```
Comment (craftworkgames) [inline Source/Demos/Demo.Particles/Game1.cs]: Just curious, why do you put the window at position zero? Is the default position a problem on Mac's?

Comment (LithiumToast) [inline Source/Demos/Demo.Particles/Game1.cs]: I'll double check by upgrading parallels, but I think there is problems with retina screens and MonoGame. I know prime31 was having issues with retina screens and MonoGame as well.

File: Source/MonoGame.Extended/Graphics/PrimitiveTypeExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+namespace MonoGame.Extended.Graphics
+{
+    public static class PrimitiveTypeExtensions
+    {
+        public static int GetPrimitiveCount(this PrimitiveType primitiveType, int verticesOrIndicesCount)
+        {
+            switch (primitiveType)
+            {
+                case PrimitiveType.LineStrip:
+                {
```
Comment (craftworkgames) [inline Source/MonoGame.Extended/Graphics/PrimitiveTypeExtensions.cs]: Putting these extra { 's in the case statements takes up a lot of extra lines for no reason. I know it's a personal preference. I'm just not a fan of having extra code that doesn't do anything soley for layout purposes.

Comment (LithiumToast) [inline Source/MonoGame.Extended/Graphics/PrimitiveTypeExtensions.cs]: Here they can be removed no problem. It's just when each case statement has a fair bit of code that it's easier, in my opinion, to read. The curly braces also create a new stack frame which means I can use the same variable names inside each case statement without clashing. Again, really only useful when there is a fair bit of code per case statement.

**Comment by craftworkgames** [comment]
Overall I really like your work. There's not really much to say about it. The code quality is excellent.

I made a couple of minor comments, but there's nothing actually wrong with it. I'd be happy to pull it whenever you're ready. Obviously there's some methods not implemented but I trust you'll get that done when you're ready.

I haven't actually tested it yet. I believe github allows a PR to be pulled locally before pulling into the `develop` branch, so I'll probably do that later.

File: Source/MonoGame.Extended/Graphics/PrimitiveTypeExtensions.cs (hunk 2 above)
Comment (craftworkgames) [inline Source/MonoGame.Extended/Graphics/PrimitiveTypeExtensions.cs]: Yeah, fair enough. I can't expect everyone to have exactly the same coding style. Overall I like the way you do things. I just wanted to mention it so you know I'm actually doing a code review ;)

File: Source/Demos/Demo.Particles/Game1.cs (hunk 1 above)
Comment (craftworkgames) [inline Source/Demos/Demo.Particles/Game1.cs]: Okay good to know. Having the window in the top left corner feels a little odd but obviously it's not really doing any harm. It is what it is I guess.

**Comment by LithiumToast** [comment]
@craftworkgames The only code that is currently filled in is the `UserPrimitivesBatchDrawer`. I still need to fill in the `DynamicVertexBatchDrawer`, `ImmediateBatchQueuer`, `DeferredBatchQueuer` and `DrawContextBatchQueuer`. The bare minimum to test the code to actually draw something would need the `ImmediateBatchQueuer` filled in so `BatchSortMode.Immediate` and `BatchDrawStrategy.UserPrimitives` can be used to actually see something on screen. `ImmediateBatchQueuer` is the easiest to implement (it's dead simple straightforward) and I will probably fill that class in tomorrow.

**Comment by craftworkgames** [comment]
Sweet.

**Comment by LithiumToast** [comment]
Going to finish up the rest of the deferred batch queuer tomorrow and test it.

**Comment by craftworkgames** [comment]
Great :)

**Comment by LithiumToast** [comment]
I created a new `SpriteBatch` class using `PrimitiveBatch`; it seems to be working fine. Nuclex code doesn't have the implementation for `DrawContextQueuer` so I'll have to think about that.

File: Source/Demos/Demo.PrimitiveBatch/Game1.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
+﻿using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using Microsoft.Xna.Framework.Input;
+using MonoGame.Extended.Graphics;
+using MonoGame.Extended.Graphics.Batching;
+using SpriteBatch = MonoGame.Extended.Graphics.Batching.SpriteBatch;
```
Comment (craftworkgames) [inline Source/Demos/Demo.PrimitiveBatch/Game1.cs]: Hmm.. I don't think it's a good idea to call this new thing a `SpriteBatch`. It's going to get very confusing when it only differs by namespace. I'd rather prefix or suffix the name with another word to make it really clear you're using an alternative to the normal sprite batch.

Comment (LithiumToast) [inline Source/Demos/Demo.PrimitiveBatch/Game1.cs]: Ok, any suggestions?

Comment (craftworkgames) [inline Source/Demos/Demo.PrimitiveBatch/Game1.cs]: `PrimitiveSpriteBatch` maybe. Or `SpriteBatchExtended`. I dunno, what would you say is the primary thing it does different to the normal sprite batch?

Comment (LithiumToast) [inline Source/Demos/Demo.PrimitiveBatch/Game1.cs]: `Vector3` as position (depth is no longer a parameter); I will add ability to pass 1 or 4 `Color` structs to use for the 4 corners of the quad; consistent API to `PrimitiveBatch` for choosing a `BatchDrawStrategy`, maximum batch size, and using a `IDrawContext` with a `BatchSortMode` for switching and batching by state (blending, sampling, textures, depth stencil, rasterization) / effects (vertex shaders, pixel shaders).

Comment (craftworkgames) [inline Source/Demos/Demo.PrimitiveBatch/Game1.cs]: I leave it in your hands to rename it. I don't think we can leave it called `SpriteBatch` though, that's sure to confuse someone.

**Comment by LithiumToast** [comment]
I scrapped the `SpriteBatch` class in favour of extensions.

e.g.

```
primitiveBatch.Begin(BatchSortMode.Immediate);
primitiveBatch.DrawSprite(_texture, Vector3.Zero);
primitiveBatch.End();
```

**Comment by LithiumToast** [comment]
Added `SpriteColor` structure; a 4 colour structure to specify the corner colours of a sprite.

I'm having issues with `BasicEffect` for any colours but white and black. I'll probably have to include a custom pixel shader into the demo.

**Comment by craftworkgames** [comment]
Good idea.

**Comment by LithiumToast** [comment]
Dropped the idea of different sort modes for deferred rendering; sorting is currently always enabled and a sort key can be passed with `PrimitiveBatch.Draw()`. Dropped `IDrawContext`; using `Effect` instead passing it to `PrimitiveBatch.Begin()` instead of `PrimitiveBatch.Draw()`.

**Comment by LithiumToast** [comment]
So `BitVector32` was not available in the PCL profile, so I created the struct along with unit tests.
Going to use `BitVector32` as the sorting key for deferred rendering. I added back `IDrawContext` meanwhile while I figure out the API to get sorted draw calls for textures and depth.

**Comment by LithiumToast** [comment]
So I think I figured out how to sort draw calls and state changes for deferred rendering using struct union:

e.g.

```
 [StructLayout(LayoutKind.Explicit, Pack = 1)]
    internal struct BatchOperation
    {
        [FieldOffset(4)]
        internal BatchDrawOperation DrawOperation;

        [FieldOffset(4)]
        internal BatchStateOperation StateOperation;
    }
```

The idea is to place a change of rendering state, e.g. switching textures or blending state, between draw calls so state changes can be minimized.
This will be done by having an array of these `BatchOperation` structs which is sorted and iterated when the `PrimitiveBatch` is flushed.

**Comment by craftworkgames** [comment]
Very cool. You're getting into areas I haven't had a lot of experience with. I'm excited to see what you come up with.

**Comment by LithiumToast** [comment]
Well that was a dead end. You can't overlap references and values in C# struct union; I wanted to have a delegate to change the render state. Could just use an index instead but then `PrimitiveBatch` needs a `Dictionary` to get / set the delegate by key... I'm just going to have one struct to share stuff for drawing and switching render states.

**Comment by LithiumToast** [comment]
Added `Material` class to represent a render state (Blend, DepthStencil, etc) for a draw call. I need to create a vertex / pixel shader for sprites; `BasicEffect` is not the greatest.

**Comment by LithiumToast** [comment]
@craftworkgames I think this is done for now. I'll need to create another demo for a game though to use it in a game scenario.

**Comment by craftworkgames** [comment]
I just tried running the demo and got

    An unhandled exception of type 'Microsoft.Xna.Framework.Content.ContentLoadException' occurred in MonoGame.Framework.dll

    Additional information: Could not load PolygonEffect asset as a non-content file!

**Comment by LithiumToast** [comment]
@craftworkgames That item should no longer exist. Did you base of the latest commit? I'll double check by doing a clone of the branch.

**Comment by LithiumToast** [comment]
@craftworkgames OK sorry about, it's fixed now.

**Comment by craftworkgames** [comment]
Merging this into `develop` is a bit of a nightmare.

**Comment by LithiumToast** [comment]
@craftworkgames Pulled develop branch in.

**Comment by craftworkgames** [comment]
Sorry it took so long to merge this. I've been sick this past week.

**Comment by LithiumToast** [comment]
Build server is complaining that it can't build the effect for the demo:

```
C:/TeamCity/buildAgent/work/982dbc876ae2b3b0/Source/Demos/Demo.PrimitiveBatch/Content/SpriteEffect.fx Processor 'EffectProcessor' had unexpected failure! 
C:/TeamCity/buildAgent/work/982dbc876ae2b3b0/Source/Demos/Demo.PrimitiveBatch/Content/PrimitiveEffect.fx Processor 'EffectProcessor' had unexpected failure! 
C:\Program Files (x86)\MSBuild\MonoGame\v3.0\MonoGame.Content.Builder.targets(89, 5): error MSB3073: The command ""C:\Program Files (x86)\MSBuild\MonoGame\v3.0\Tools\MGCB.exe" /@:"C:\TeamCity\buildAgent\work\982dbc876ae2b3b0\Source\Demos\Demo.PrimitiveBatch\Content\Content.mgcb" /platform:DesktopGL /outputDir:"C:\TeamCity\buildAgent\work\982dbc876ae2b3b0\Source\Demos\Demo.PrimitiveBatch\Content\bin\DesktopGL" /intermediateDir:"C:\TeamCity\buildAgent\work\982dbc876ae2b3b0\Source\Demos\Demo.PrimitiveBatch\Content\obj\DesktopGL" /quiet" exited with code 2.
```

**Comment by craftworkgames** [comment]
The issues with the build server turned out to be this:

    System.DllNotFoundException: Unable to load DLL 'd3dcompiler_43.dll': The specified module could not be found.

Apparently you need that DLL to compile the shaders and the Azure machine we host the build server on didn't have it. I found [a post](http://stackoverflow.com/questions/13544367/cant-find-d3dcompiler-43-dll-for-directx-programming) on stack overflow that helped. I mostly had to figure it out from the comments on the answers though and install a bunch of things like the Windows SDK / DirectX, etc. Eventually it worked :)

**Comment by LithiumToast** [comment]
Great!

**Pull request merged**